tracker/dmlc_tracker/local.py

```diff
@@ -13,28 +13,42 @@ def exec_cmd(cmd, role, taskid, pass_env):
     """Execute the command line command."""
     if cmd[0].find('/') == -1 and os.path.exists(cmd[0]) and os.name != 'nt':
         cmd[0] = './' + cmd[0]
-    cmd = ' '.join(cmd)
+    cmdline = ' '.join(cmd)
     env = os.environ.copy()
     for k, v in pass_env.items():
         env[k] = str(v)
 
     env['DMLC_TASK_ID'] = str(taskid)
     env['DMLC_ROLE'] = role
     env['DMLC_JOB_CLUSTER'] = 'local'
-
     num_retry = env.get('DMLC_NUM_ATTEMPT', 0)
 
+    #overwrite default num of retry with commandline value
+    for param in cmd:
+        if param.startswith('DMLC_NUM_ATTEMPT'):
+            num_retry = int(param.split('=')[1])
+    logging.debug('num of retry %d',num_retry)
+
     while True:
         if os.name == 'nt':
-            ret = subprocess.call(cmd, shell=True, env=env)
+            ret = subprocess.call(cmdline, shell=True, env=env)
         else:
-            ret = subprocess.call(cmd, shell=True, executable='bash', env=env)
+            ret = subprocess.call(cmdline, shell=True, executable='bash', env=env)
         if ret == 0:
             logging.debug('Thread %d exit with 0', taskid)
             return
         else:
             num_retry -= 1
+            newcmd = []
             if num_retry >= 0:
+                # failure trail increase by 1 and restart failed worker
+                for arg in cmd:
+                    if arg.startswith('rabit_num_trial'):
```

Comment:
I might have missed this here, I don't think we should leak more rabit stuff in dmlc-core (hopefully we should rename RabitTracker), essentially they are two different projects

Reply:
Described on PR item 3, the semantic of rabit_num_trial is to capture the number of worker failures caused by mock. v.s DMLC_NUM_ATTEMPT is setting number of retries in total local tracker can do. The issue really some how up to date rabit_num_trial needs to feed back into rabit init in order to make rabit behavior as expected. Otherwise it will be infinite crash...

https://github.com/dmlc/rabit/blob/master/src/allreduce_mock.h#L164

One previous attempt was to reuse DMLC_NUM_ATTEMPT as replacement of rabit_num_trial. The problem is rabit_num_trial starts with 0 and DMLC_NUM_ATTEMPT needs to >= number of crashes for tests to pass. That basically broke tests in rabit.

Reply:
ok, in this case, why we need to pass a value as `rabit_num_trial`?

we should remove it from both of dmlc-core and rabit, right?

Reply:
how can restarted rabit mock worker know which **trail** it is running and decide if it will crash or not. There needs something pass via arguments...
https://github.com/dmlc/rabit/blob/master/src/allreduce_mock.h#L52

Reply:
I would suggest add another argument called 
DMLC_WORKER_MAX_RETRY along with DMLC_NUM_ATTEMPT and remove rabit_num_trial from core.

Reply:
just DMLC_MAX_RETRY and consume rabit_num_trial from rabit

Reply:
https://github.com/dmlc/dmlc-core/pull/512/files